File: reference/src/adapters/scanner.ts
/**
 * Local filesystem adapter scanner.
 * Reads adapter metadata.yaml files from a directory structure.
 */

import { readdir, readFile } from 'node:fs/promises';
import { join } from 'node:path';
import { parse as parseYaml } from 'yaml';
import type { AdapterMetadata } from './registry.js';
import { validateAdapterMetadata } from './registry.js';

export interface ScanOptions {
  /** Base directory containing adapter subdirectories. */
  basePath: string;
  /** Whether to skip adapters with invalid metadata. Default: true. */
  skipInvalid?: boolean;
}

export interface ScanResult {
  adapters: AdapterMetadata[];
  errors: Array<{ path: string; error: string }>;
}

/**
 * Parse a metadata YAML string into AdapterMetadata.
 */
export function parseMetadataYaml(content: string): AdapterMetadata {
  const parsed = parseYaml(content);
  if (!parsed || typeof parsed !== 'object') {
    throw new Error('Invalid YAML: expected an object');
  }
  return parsed as AdapterMetadata;
}

/**
 * Scan a directory for adapter metadata.yaml files.
 * Expects structure: `<basePath>/<adapter-name>/metadata.yaml`
 */
export async function scanLocalAdapters(options: ScanOptions): Promise<ScanResult> {
  const { basePath, skipInvalid = true } = options;
  const adapters: AdapterMetadata[] = [];
  const errors: ScanResult['errors'] = [];

  let entries: string[];
  try {
    const dirEntries = await readdir(basePath, { withFileTypes: true });
    entries = dirEntries.filter((e) => e.isDirectory()).map((e) => e.name);
  } catch (err) {
    return {
      adapters: [],
      errors: [
        {
          path: basePath,
          error: `Cannot read directory: ${err instanceof Error ? err.message : String(err)}`,
        },
      ],
    };
  }

  for (const dirName of entries) {
    const metadataPath = join(basePath, dirName, 'metadata.yaml');
    try {
      const content = await readFile(metadataPath, 'utf-8');
      const metadata = parseMetadataYaml(content);
      const validation = validateAdapterMetadata(metadata);
      if (validation.valid) {
        adapters.push(metadata);
      } else if (!skipInvalid) {
        errors.push({ path: metadataPath, error: validation.errors.join('; ') });
      } else {
        errors.push({ path: metadataPath, error: validation.errors.join('; ') });
      }
    } catch (err) {
      errors.push({
        path: metadataPath,
        error: err instanceof Error ? err.message : String(err),
      });
    }
  }

  return { adapters, errors };
}

File: sdk-python/src/ai_sdlc/adapters/scanner.py
"""Local filesystem adapter scanner.

Reads adapter metadata.yaml files from a directory structure.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from pathlib import Path

import yaml

from ai_sdlc.adapters.registry import AdapterMetadata, validate_adapter_metadata


@dataclass
class ScanOptions:
    base_path: str
    skip_invalid: bool = True


@dataclass
class ScanError:
    path: str
    error: str


@dataclass
class ScanResult:
    adapters: list[AdapterMetadata] = field(default_factory=list)
    errors: list[ScanError] = field(default_factory=list)


def parse_metadata_yaml(content: str) -> AdapterMetadata:
    """Parse a metadata YAML string into AdapterMetadata."""
    parsed = yaml.safe_load(content)
    if not parsed or not isinstance(parsed, dict):
        raise ValueError("Invalid YAML: expected an object")
    return AdapterMetadata(
        name=parsed.get("name", ""),
        display_name=parsed.get("displayName", ""),
        description=parsed.get("description", ""),
        version=parsed.get("version", ""),
        stability=parsed.get("stability", "alpha"),
        interfaces=parsed.get("interfaces", []),
        owner=parsed.get("owner", ""),
        spec_versions=parsed.get("specVersions", []),
        repository=parsed.get("repository"),
        dependencies=parsed.get("dependencies"),
    )


async def scan_local_adapters(options: ScanOptions) -> ScanResult:
    """Scan a directory for adapter metadata.yaml files.

    Expects structure: ``<base_path>/<adapter-name>/metadata.yaml``
    """
    base = Path(options.base_path)
    adapters: list[AdapterMetadata] = []
    errors: list[ScanError] = []

    if not base.is_dir():
        return ScanResult(
            errors=[ScanError(path=str(base), error=f"Cannot read directory: {base}")]
        )

    for entry in sorted(base.iterdir()):
        if not entry.is_dir():
            continue
        metadata_path = entry / "metadata.yaml"
        try:
            content = metadata_path.read_text(encoding="utf-8")
            metadata = parse_metadata_yaml(content)
            validation = validate_adapter_metadata(metadata)
            if validation.valid:
                adapters.append(metadata)
            else:
                errors.append(
                    ScanError(path=str(metadata_path), error="; ".join(validation.errors))
                )
        except Exception as err:
            errors.append(ScanError(path=str(metadata_path), error=str(err)))

    return ScanResult(adapters=adapters, errors=errors)
